dcicutils/lang_utils.py

```diff
@@ -50,18 +52,86 @@ def _special_case_plural(cls, word: str) -> str:
+    # in that the plural of 'a variant referencing a gene' would be 'variants referencing genes', just as
+    # the plural of 'a gene referenced by a variant' would be 'genes referenced by variants'.
+
+    _COMPOUND_PLURAL_PSEUDO_CONNECTIVES = [
+        'using', 'containing', 'including', 'referencing', 'naming',
+        'used by', 'contained by', 'included by', 'referenced by', 'named by',
+        'used in', 'contained in', 'included in', 'referenced in', 'named in',
+    ]
+
+    _COMPOUND_PLURAL_PREPOSITIONS = _COMPOUND_PLURAL_SIMPLE_PREPOSITIONS + _COMPOUND_PLURAL_PSEUDO_CONNECTIVES
+
+    _NOUN_WITH_PREPOSITIONAL_ATTACHMENT = re.compile(
+        # Note use of *? to get minimal match in group1, so that we'll find the first preposition, not a later one
+        # Specifically, the use of "*?" makes us treat multiple prepositions like a-of-b-of-c by recursing as
+        # (a)(-)(of)(-)(b-of-c). If we used "*" instead of "*?", we would get (a-of-b)(-)(of)(-)(c).
+        # If we were only doing hyphenated items, it wouldn't matter a whole lot.
+        # But in the words part, it matters a great deal because with spaces and hyphens interleaved, there is a
+        # risk of parsing "a-of-b of c-of-d" as (a-of-b of c)(-)(of)(-)(d), whereas I think (a-of-b)( )(of)( )(c-of-d)
+        # is more likely to get a better parse. -kmp 29-Aug-2021
+        f"""^(?:
+                 # Either these 5 will match (and the next 5 will not)
+                 # to ONLY match a hyphenated compound word like son-in-law as (son)(-)(in)(-)(law)
+                 ([a-z][a-z-]*?)  # shortest possible (i.e., first) block of hyphenated words preceding "-<prep>-"
+                 ([-])            # pre-preposition-hyphen to make sure it's part of the same compound.
+                 ({'|'.join(_COMPOUND_PLURAL_PREPOSITIONS).replace(' ', '[ ]')}) # matches the prep, as (about|at|...)
+                 ([-])            # hyphen on the other side of prep
+                 ([a-z-]*)        # we're less fussy about this. It could contain more preps, for example.
+              |
+                 # or these 5 will match (and the previous 5 will not)...
+                 # to match shortest (i.e., first) words (included hyphenated words) leading to " <prep> "
+                 # so 'the son-in-law of the proband' becomes (the son-in-law)( )(of)( )(the proband)
+                 ([a-z]        # first token starts with an alphabetic,
+                  [a-z- ]*?    # matches any number of words, which may be hyphenated,
+                  [a-z])       # and ends in an alphabetic
+                 ([ ]+)        # unlike with hyphenation, any number of pre-<prep> spaces is ok
+                 ({'|'.join(_COMPOUND_PLURAL_PREPOSITIONS).replace(' ', '[ ]')}) # matches the prep, as (about|at|...)
+                 ([ ]+)        # any amount of whitespace on the other side
+                 (.*)          # anything else that follows first space-delimeted preposition
```

Comment:
Nice little regex refresher 😄   - I am taking what you say here at face value and not double checking it, just FYI as this is not one of my stronger areas.

Reply:
the general approach is something i've been refining for years in various environments so i have a general theory of what i'm trying to do, but i'm doing the implementation from scratch here just from the ideas, so some risk of error. And I've never tried to do it with regexps.

still, the risk of being wrong is not high. the unit test coverage is decent, but i could probably fix bugs if they came up.